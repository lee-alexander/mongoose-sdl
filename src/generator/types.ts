import { DbDefinition, Enum, Schema, SchemaDataType, Model } from '../types';
import { assertUnreachable } from '../util';

export function generateTypes(sdl: DbDefinition): string {
  const enums = Object.entries(sdl.enums).map(([name, data]) => generateEnumType(name, data));
  const models = Object.entries(sdl.models).map(([name, data]) => generateModelType(getModelTypeName(name), data));
  const schemas = Object.entries(sdl.schemas).map(([name, data]) => generateSchemaType(getSchemaTypeName(name), data));

  return [`import { Types, Document } from 'mongoose';`, ...enums, ...models, ...schemas].join('\n\n');
}

function generateEnumType(name: string, data: Enum) {
  const enumValues = data.values.map((value) => `  ${value} = '${value}',`).join('\n');
  return `export enum ${name} {\n${enumValues}\n}`;
}

function generateSchemaType(name: string, data: Schema) {
  const defaultFields = ['  id: string;', '  createdAt: Date;', '  updatedAt: Date;'].join('\n');
  const customFields = Object.entries(data)
    .map(
      ([name, data]) =>
        `  ${data.isImmutable ? 'readonly ' : ''}${name}: ${getTypeName(data.dataType)}${
          data.isRequired ? '' : ' | null | undefined'
        };`
    )
    .join('\n');

  return `export interface ${name} extends Document {\n${defaultFields}\n${customFields}\n}`;
}

function generateModelType(name: string, model: Model) {
  return generateSchemaType(name, model.schema);
}

function getTypeName(field: SchemaDataType): string {
  switch (field.type) {
    case 'String':
      return 'string';
    case 'Boolean':
      return 'boolean';
    case 'Number':
      return 'number';
    case 'Date':
      return 'Date';
    case 'Enum':
      return field.refEnum;
    case 'Schema':
      return getSchemaTypeName(field.refSchema);
    case 'ObjectId':
      return 'Types.ObjectId';
    case 'Array': {
      const elementType = getTypeName(field.elementType);
      return field.elementRequired ? `${elementType}[]` : `(${elementType} | null)[]`;
    }
    case 'Map': {
      const elementType = getTypeName(field.elementType);
      return `Map<string, ${elementType}${field.elementRequired ? '' : ' | null'}>`;
    }
    default:
      assertUnreachable(field);
  }
}

export function getSchemaTypeName(name: string) {
  return `${name}SchemaDocument`;
}

export function getModelTypeName(name: string) {
  return `${name}Document`;
}
