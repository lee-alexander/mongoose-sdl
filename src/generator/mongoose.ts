import { DbDefinition, Schema, SchemaField, SchemaDataTypeWithoutArray, Model } from '../types';
import { assertUnreachable, notNullOrUndefined } from '../util';
import { getModelTypeName, getSchemaTypeName } from './types';

export function generateMongoose(sdl: DbDefinition): string {
  // TODO - DAG ordering for schemas by dependencies
  // TODO handle recursive schemas

  const schemas = Object.entries(sdl.schemas).map(([name, data]) =>
    generateSchema(getSchemaName(name), getSchemaTypeName(name), data)
  );
  const models = Object.entries(sdl.models).map(([name, data]) => generateModel(name, data));

  return [`import { Schema, model } from 'mongoose';`, ...schemas, ...models].join('\n');
}

function generateModel(name: string, model: Model) {
  const schemaContent = generateSchema(getSchemaName(name), getModelTypeName(name), model.schema);
  const modelContent = `export const ${getModelName(name)} = model<${getModelTypeName(
    name
  )}>('${name}', ${getSchemaName(name)});`;
  return `${schemaContent}\n\n${modelContent}\n`;
}

function generateSchema(name: string, typeName: string, data: Schema) {
  return `const ${name} = new Schema<${typeName}>(
  {\n${Object.entries(data)
    .map(([name, data]) => `    ${name}: ${getSchemaFieldDefinition(data)},`)
    .join('\n')}
  },
  { timestamps: true }
)`;
}

function getSchemaFieldDefinition(field: SchemaField): string {
  const baseDefinition = [
    field.isRequired ? 'required: true' : null,
    field.isIndex ? 'index: true' : null,
    field.isUnique ? 'unique: true' : null,
  ];

  let customDefinition: string;
  if (field.dataType.type === 'Array') {
    const childType = getSchemaFieldDataDefinition(field.dataType.elementType);
    const childFields = [childType, field.dataType.elementRequired ? 'required: true' : ''].join(', ');
    customDefinition = `type: [{ ${childFields} }]`;
  } else {
    customDefinition = getSchemaFieldDataDefinition(field.dataType);
  }

  return `{ ${[...baseDefinition, customDefinition].filter(notNullOrUndefined).join(', ')} }`;
}

function getSchemaFieldDataDefinition(data: SchemaDataTypeWithoutArray): string {
  switch (data.type) {
    case 'String':
    case 'Boolean':
    case 'Number':
    case 'Date':
      return `type: ${data.type}`;
    case 'Enum':
      return `type: String, enum: ${data.refEnum}`;
    case 'ObjectId':
      return `type: Schema.Types.ObjectId, ref: '${data.refModel}'`;
    case 'Schema':
      return `type: ${getSchemaName(data.refSchema)}`;
    default:
      assertUnreachable(data);
  }
}

function getSchemaName(name: string) {
  return `${name}Schema`;
}

function getModelName(name: string) {
  return `${name}Model`;
}
