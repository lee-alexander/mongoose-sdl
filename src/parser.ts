import { promises as fs } from 'fs';
import { groupItemsBy, toDictionary, uniqueValues, unwrap } from './util';
import { Enum, Schema, Model, DbDefinition, FlatSchemaDataType } from './types';

// Break overall document down into key sections - enums, models, schemas
const TopLevelRegex = /(enum|model|schema) ([^{]*) {([^}]*)}/g;

export async function parseDbDefinitionFile(path: string): Promise<DbDefinition> {
  const fileBuffer = await fs.readFile(path);
  const contents = fileBuffer.toString();
  const parsedContent = parseContent(contents, TopLevelRegex);

  const namesWithType = parsedContent.map(([, type, name]) => `${name}${type}`);
  if (namesWithType.length !== uniqueValues(namesWithType).length) {
    throw new Error('Duplicate type name detected');
  }

  const contentByType = groupItemsBy(
    parsedContent,
    ([, type]) => type as 'enum' | 'model' | 'schema',
    ([, , name, contents]) => ({ name, contents })
  );

  const namedTypes: NamedTypes = {
    enums: new Set((contentByType['enum'] ?? []).map((d) => d.name)),
    models: new Set((contentByType['model'] ?? []).map((d) => d.name)),
    schemas: new Set((contentByType['schema'] ?? []).map((d) => d.name)),
  };

  const enums = toDictionary(
    contentByType['enum'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseEnumContents(contents)
  );

  const schemas = toDictionary(
    contentByType['schema'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseSchemaContents(contents, namedTypes)
  );

  const models = toDictionary(
    contentByType['model'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseModelContents(contents, namedTypes)
  );

  return { enums, schemas, models };
}

const EnumRegex = /(\w+)/g;
function parseEnumContents(contents: string): Enum {
  const values = parseContent(contents, EnumRegex);

  return {
    values: values.map(([value]) => value),
  };
}

// Break down "key: type @directive1 @directive2" lines into
// (1) key
// (2) array element type, if array
// (3) array element type is required (!), if array
// (4) map element type, if map
// (5) map element type is required (!), if map
// (6) type, if not array or map
// (7) type is required (!)
// (8) directives
const KeyValueDirectiveRegex = /(\w+): (?:\[(\w+)(!?)\]|Map<(\w+)(!?)>|(\w+))(!?)([^\n]*)/g;

// Break down @directive1 @directive2 into (1) directive1 (2) directive2
const DirectiveRegex = /@(\w+)/g;

function parseSchemaContents(contents: string, namedTypes: NamedTypes): Schema {
  const parsedContent = parseContent(contents, KeyValueDirectiveRegex);
  const parsedFields = parsedContent.map(
    ([
      ,
      fieldName,
      arrayElementType,
      arrayElementRequired,
      mapElementType,
      mapElementRequired,
      nonArrayType,
      required,
      directives,
    ]) => ({
      fieldName,
      fieldType: unwrap(nonArrayType || arrayElementType || mapElementType),
      isArray: Boolean(arrayElementType),
      isArrayElementRequired: arrayElementRequired === '!',
      isMap: Boolean(mapElementType),
      isMapElementRequired: mapElementRequired === '!',
      isRequired: required === '!',
      directives: directives ? parseContent(directives, DirectiveRegex).map(([, directive]) => directive) : [],
    })
  );

  const invalidDirectives = parsedFields
    .flatMap((f) => f.directives)
    .filter((d) => d !== 'index' && d !== 'unique' && d !== 'immutable');
  if (invalidDirectives.length > 0) {
    throw new Error('Unknown directives: ' + invalidDirectives);
  }

  return toDictionary(
    parsedFields,
    (f) => f.fieldName,
    (f) => ({
      dataType: f.isArray
        ? {
            type: 'Array' as const,
            elementType: parseDataType(f.fieldName, f.fieldType, namedTypes),
            elementRequired: f.isArrayElementRequired,
          }
        : f.isMap
        ? {
            type: 'Map' as const,
            elementType: parseDataType(f.fieldName, f.fieldType, namedTypes),
            elementRequired: f.isMapElementRequired,
          }
        : parseDataType(f.fieldName, f.fieldType, namedTypes),
      isRequired: f.isRequired,
      isIndex: f.directives.includes('index'),
      isUnique: f.directives.includes('unique'),
      isImmutable: f.directives.includes('immutable'),
    })
  );
}

function parseModelContents(contents: string, namedTypes: NamedTypes): Model {
  return {
    schema: parseSchemaContents(contents, namedTypes),
  };
}

function parseDataType(fieldName: string, fieldType: string, namedTypes: NamedTypes): FlatSchemaDataType {
  if (fieldType === 'String' || fieldType === 'Boolean' || fieldType === 'Number' || fieldType === 'Date') {
    return { type: fieldType };
  }

  if (namedTypes.enums.has(fieldType)) {
    return {
      type: 'Enum',
      refEnum: fieldType,
    };
  }

  if (namedTypes.models.has(fieldType)) {
    return {
      type: 'ObjectId',
      refModel: fieldType,
    };
  }

  if (namedTypes.schemas.has(fieldType)) {
    return {
      type: 'Schema',
      refSchema: fieldType,
    };
  }

  throw new Error(`Field ${fieldName} has unknown type ${fieldType}`);
}

interface NamedTypes {
  enums: Set<string>;
  schemas: Set<string>;
  models: Set<string>;
}

function parseContent(content: string, regex: RegExp) {
  const result = content.matchAll(regex);
  const unmatched = content.replaceAll(regex, '').trim();
  if (unmatched.length > 0) {
    throw new Error('Unexpected syntax near:\n' + unmatched);
  }
  return [...result];
}
