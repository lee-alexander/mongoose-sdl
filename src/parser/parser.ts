import { promises as fs } from 'fs';
import { groupItemsBy, toDictionary, uniqueValues, unwrap } from '../util';
import { Enum, Schema, Model, DbDefinition, FlatSchemaDataType, Union, Unions } from '../types';

// Break overall document down into key sections - enums, models, schemas
const TopLevelRegex = /(?:(enum|model|schema) ([^{]*) {([^}]*)})|(?:external (\w*))|(?:union (\w*) = ([\w |]*))/g;

export async function parseDbDefinitionFile(path: string): Promise<DbDefinition> {
  const fileBuffer = await fs.readFile(path);
  const contents = fileBuffer.toString();
  const parsedContent = parseContent(contents, TopLevelRegex);

  const topLevelNames = parsedContent.map(
    ([, , name, _, externalTypeName, unionName]) => name || externalTypeName || unionName
  );
  if (topLevelNames.length !== uniqueValues(topLevelNames).length) {
    throw new Error('Duplicate schema/model/enum/external/union name detected');
  }

  const contentByType = groupItemsBy(
    parsedContent,
    ([, type, , , externalTypeName, unionName]) =>
      externalTypeName ? 'external' : unionName ? 'union' : (type as 'enum' | 'model' | 'schema'),
    ([, , name, contents, externalTypeName, unionName, unionContents]) => ({
      name: name || externalTypeName || unionName,
      contents: contents || unionContents || '',
    })
  );

  const namedTypes: NamedTypes = {
    enums: new Set((contentByType['enum'] ?? []).map((d) => d.name)),
    models: new Set((contentByType['model'] ?? []).map((d) => d.name)),
    schemas: new Set((contentByType['schema'] ?? []).map((d) => d.name)),
    externals: new Set((contentByType['external'] ?? []).map((d) => d.name)),
    unions: new Set((contentByType['union'] ?? []).map((d) => d.name)),
  };

  const unions = toDictionary(
    contentByType['union'] ?? [],
    ({ name }) => name,
    ({ name, contents }) => parseUnionContents(name, contents, namedTypes)
  );

  const enums = toDictionary(
    contentByType['enum'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseEnumContents(contents)
  );

  const schemas = toDictionary(
    contentByType['schema'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseSchemaContents(contents, namedTypes, unions)
  );

  const models = toDictionary(
    contentByType['model'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseModelContents(contents, namedTypes, unions)
  );

  const externals = (contentByType['external'] ?? []).map((c) => c.name);

  return { enums, schemas, models, unions, externals };
}

const EnumRegex = /(\w+)/g;
function parseEnumContents(contents: string): Enum {
  const values = parseContent(contents, EnumRegex);

  return {
    values: values.map(([value]) => value),
  };
}

function parseUnionContents(name: string, contents: string, namedTypes: NamedTypes): Union {
  const refModels = contents.split('|').map((m) => m.trim());
  const unknownModels = refModels.filter((model) => !namedTypes.models.has(model));
  if (unknownModels.length > 0) {
    throw new Error(`Unknown model names in union ${name}: ${unknownModels.join(', ')}`);
  }
  return { refModels };
}

// Break down "key: type @directive1 @directive2" lines into
// (1) key
// (2) array element type, if array
// (3) array element type is required (!), if array
// (4) map element type, if map
// (5) map element type is required (!), if map
// (6) type, if not array or map
// (7) type is required (!)
// (8) directives
const KeyValueDirectiveRegex = /(\w+): (?:\[(\w+)(!?)\]|Map<(\w+)(!?)>|(\w+))(!?)([^\n]*)/g;

// Break down @directive1 @directive2 into (1) directive1 (2) directive2
const DirectiveRegex = /@(\w+)/g;

function parseSchemaContents(contents: string, namedTypes: NamedTypes, unions: Unions): Schema {
  const parsedContent = parseContent(contents, KeyValueDirectiveRegex);
  const parsedFields = parsedContent.map(
    ([
      ,
      fieldName,
      arrayElementType,
      arrayElementRequired,
      mapElementType,
      mapElementRequired,
      nonArrayType,
      required,
      directives,
    ]) => ({
      fieldName,
      fieldType: unwrap(nonArrayType || arrayElementType || mapElementType),
      isArray: Boolean(arrayElementType),
      isArrayElementRequired: arrayElementRequired === '!',
      isMap: Boolean(mapElementType),
      isMapElementRequired: mapElementRequired === '!',
      isRequired: required === '!',
      directives: directives ? parseContent(directives, DirectiveRegex).map(([, directive]) => directive) : [],
    })
  );

  const invalidDirectives = parsedFields
    .flatMap((f) => f.directives)
    .filter((d) => d !== 'index' && d !== 'unique' && d !== 'immutable' && d !== 'virtual' && d !== 'validate');
  if (invalidDirectives.length > 0) {
    throw new Error('Unknown directives: ' + invalidDirectives);
  }

  return toDictionary(
    parsedFields,
    (f) => f.fieldName,
    (f) => ({
      dataType: f.isArray
        ? {
            type: 'Array' as const,
            elementType: parseDataType(f.fieldName, f.fieldType, namedTypes, unions),
            elementRequired: f.isArrayElementRequired,
          }
        : f.isMap
        ? {
            type: 'Map' as const,
            elementType: parseDataType(f.fieldName, f.fieldType, namedTypes, unions),
            elementRequired: f.isMapElementRequired,
          }
        : parseDataType(f.fieldName, f.fieldType, namedTypes, unions),
      isRequired: f.isRequired,
      isIndex: f.directives.includes('index'),
      isUnique: f.directives.includes('unique'),
      isImmutable: f.directives.includes('immutable'),
      isVirtual: f.directives.includes('virtual'),
      isValidatable: f.directives.includes('validate'),
    })
  );
}

function parseModelContents(contents: string, namedTypes: NamedTypes, unions: Unions): Model {
  return {
    schema: parseSchemaContents(contents, namedTypes, unions),
  };
}

function parseDataType(
  fieldName: string,
  fieldType: string,
  namedTypes: NamedTypes,
  unions: Unions
): FlatSchemaDataType {
  if (fieldType === 'String' || fieldType === 'Boolean' || fieldType === 'Number' || fieldType === 'Date') {
    return { type: fieldType };
  }

  if (namedTypes.enums.has(fieldType)) {
    return {
      type: 'Enum',
      refEnum: fieldType,
    };
  }

  if (fieldType === 'ObjectId') {
    return {
      type: 'ObjectId',
      refModels: [],
    };
  }

  if (namedTypes.externals.has(fieldType)) {
    return {
      type: 'External',
      refType: fieldType,
    };
  }

  if (namedTypes.models.has(fieldType)) {
    return {
      type: 'ObjectId',
      refModels: [fieldType],
    };
  }

  if (unions[fieldType]) {
    return {
      type: 'ObjectId',
      refModels: unions[fieldType].refModels,
    };
  }

  if (namedTypes.schemas.has(fieldType)) {
    return {
      type: 'Schema',
      refSchema: fieldType,
    };
  }

  throw new Error(`Field ${fieldName} has unknown type ${fieldType}`);
}

interface NamedTypes {
  enums: Set<string>;
  schemas: Set<string>;
  models: Set<string>;
  externals: Set<string>;
  unions: Set<string>;
}

function parseContent(content: string, regex: RegExp) {
  const result = content.matchAll(regex);
  const unmatched = content.replaceAll(regex, '').trim();
  if (unmatched.length > 0) {
    throw new Error('Unexpected syntax near:\n' + unmatched);
  }
  return [...result];
}
