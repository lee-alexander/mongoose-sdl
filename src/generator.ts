import { generateMongoose } from './generator/mongoose';
import { generateTypes } from './generator/types';
import { DbDefinition } from './types';

export function generate(sdl: DbDefinition): string {
  const types = generateTypes(sdl);
  const mongoose = generateMongoose(sdl);

  return `${types}\n\n${mongoose}`;
}
