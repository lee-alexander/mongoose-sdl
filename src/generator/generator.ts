import { generateMongoose } from './mongoose';
import { generateTypes } from './types';
import { DbDefinition } from '../types';

export function generate(sdl: DbDefinition): string {
  const types = generateTypes(sdl);
  const mongoose = generateMongoose(sdl);

  return `${types}\n\n${mongoose}`;
}
