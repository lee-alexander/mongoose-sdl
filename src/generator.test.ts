import { generate } from './generator';
import { parseDbDefinitionFile } from './parser';

test('generate example.mgsdl', async () => {
  const definition = await parseDbDefinitionFile('./sample-data/example.mgsdl');
  const result = generate(definition);
  expect(result).toMatchSnapshot();
});

test('generate topological-ordering.mgsdl', async () => {
  const definition = await parseDbDefinitionFile('./sample-data/topological-ordering.mgsdl');
  const result = generate(definition);
  expect(result).toMatchSnapshot();
});

test('generate cycles.mgsdl', async () => {
  const definition = await parseDbDefinitionFile('./sample-data/invalid/cycles.mgsdl');
  expect(() => generate(definition)).toThrowErrorMatchingInlineSnapshot(`"Cycle detected between schemas"`);
});
