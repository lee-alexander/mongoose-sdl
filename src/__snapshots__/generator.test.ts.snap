// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate example.mgsdl 1`] = `
"import { Types, Document } from 'mongoose';

export enum UserStatus {
  Pending = 'Pending',
  Active = 'Active',
  Inactive = 'Inactive',
}

export interface UserDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  email: string;
  name: string | null | undefined;
  status: UserStatus;
}

export interface ProjectDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  creator: Types.ObjectId;
  details: ProjectDetailsSchemaDocument;
  lastAccessedByUserId: Map<string, string | null>;
  readonly detailsByUserId: Map<string, ProjectDetailsSchemaDocument> | null | undefined;
}

export interface ProjectDetailsSchemaDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  name: string;
  alternateNames: string[];
  number: number | null | undefined;
  readonly isActive: boolean;
  startDate: Date | null | undefined;
}

import { Schema, model } from 'mongoose';

const ProjectDetailsSchema = new Schema<ProjectDetailsSchemaDocument>(
  {
    name: { required: true, type: String },
    alternateNames: { required: true, type: [{ type: String, required: true }] },
    number: { unique: true, type: Number },
    isActive: { required: true, immutable: true, type: Boolean },
    startDate: { type: Date },
  },
  { timestamps: true }
);

const UserSchema = new Schema<UserDocument>(
  {
    email: { required: true, index: true, unique: true, type: String },
    name: { type: String },
    status: { required: true, type: String, enum: UserStatus },
  },
  { timestamps: true }
);

export const UserModel = model<UserDocument>('User', UserSchema);

const ProjectSchema = new Schema<ProjectDocument>(
  {
    creator: { required: true, index: true, type: Schema.Types.ObjectId, ref: 'User' },
    details: { required: true, type: ProjectDetailsSchema },
    lastAccessedByUserId: { required: true, type: Map, of: { type: String } },
    detailsByUserId: { immutable: true, type: Map, of: { type: ProjectDetailsSchema, required: true } },
  },
  { timestamps: true }
);

export const ProjectModel = model<ProjectDocument>('Project', ProjectSchema);"
`;

exports[`generate topological-ordering.mgsdl 1`] = `
"import { Types, Document } from 'mongoose';

export interface ProjectDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  name: string;
  details: ProjectDetailsSchemaDocument;
}

export interface CompanySchemaDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  name: string | null | undefined;
}

export interface ContractorSchemaDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  name: string;
  company: CompanySchemaDocument | null | undefined;
  parentContractor: ContractorSchemaDocument;
}

export interface ProjectDetailsSchemaDocument extends Document {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  name: string;
  contractor: ContractorSchemaDocument;
}

import { Schema, model } from 'mongoose';

const CompanySchema = new Schema<CompanySchemaDocument>(
  {
    name: { type: String },
  },
  { timestamps: true }
);

const ContractorSchema = new Schema<ContractorSchemaDocument>(
  {
    name: { required: true, type: String },
    company: { type: CompanySchema },
  },
  { timestamps: true }
);
ContractorSchema.add({ parentContractor: { required: true, type: ContractorSchema } });

const ProjectDetailsSchema = new Schema<ProjectDetailsSchemaDocument>(
  {
    name: { required: true, type: String },
    contractor: { required: true, type: ContractorSchema },
  },
  { timestamps: true }
);

const ProjectSchema = new Schema<ProjectDocument>(
  {
    name: { required: true, type: String },
    details: { required: true, type: ProjectDetailsSchema },
  },
  { timestamps: true }
);

export const ProjectModel = model<ProjectDocument>('Project', ProjectSchema);"
`;
